### ISSUE TITLE
Event structure for Hardware Triggering

### ISSUE THREAD

**Issue by kidantad** [micro-manager/pycro-manager#385]
I am trying to send a TTL signal from my stage(ASI XY stage with MS 2000 controller) to my camera(HamammatsuHam_DCAM) and do a constant scan in the X direction. I used the SCAN module to control the SCAN in X. I need my camera to take 200 images when it receives the signal from the stage. Per the SCAN module, my stage has the start,end position and speed it need to do the scanning. 

I based my pycromanager code based on this example (https://pycro-manager.readthedocs.io/en/latest/application_notebooks/external_master_tutorial.html?highlight=scan#Setup-ASI-stage-control-cards-and-programmable-logic-card-in-the-Tiger-controller)

However, when I run the acquisition, it either takes a snapshot or doesn't take any image. I'm guessing that I am doing something wrong with my event structure but I don't know what. Please let me know if you have any suggestions.

For reference: I only want to move in the X axis and I need to take 200 images as I do.

**EVENT STRUCTURE**

for y in range(y_positions):  #Only 1 y position for all x positions

    for x in range(scan_positions):
        evt = {'axes': {'x' :x,'y':y} , 'x': scan_start, 'y': y_start}
        
    events.append(evt)

**POST CAMERA HOOK**

def post_camera_hook_(event,bridge,event_queue):

    # acquire core from bridge
    mmc = bridge.get_core()

    # send MS2000 command to start constant speed scan
    command='SCAN'
    mmc.set_property('XYStage','SerialCommand',command)

    return event

ready = mmc.get_property('XYStage','SerialResponse')

**ACQUISTION** 

mmc.set_property('HamamatsuHam_DCAM','TRIGGER SOURCE','EXTERNAL')
sleep(5)

with Acquisition(directory=save_dir, name=save_namprint(eventse, post_camera_hook_fn=post_camera_hook_,
                 show_display=False, max_multi_res_index=0) as acq:
    
    acq.acquire(events)

**STAGE SETUP**

command = 'SCANR X='+str(scan_start)+' Y='+str(scan_end)+' R=25'
mmc.set_property('XYStage','SerialCommand',command)

# turn off 'transmit repeated commands' for MS2000
Yes='Yes'
mmc.set_property('XYStage','OnlySendSerialCommandOnChange',Yes)

**Comment by henrypinkard** [comment]
Hard to tell from the event code. Can you print some of those events and post?

You could also try this code with the demo camera. If it acquires data as you expect then you probably have a hardware issue

**Comment by kidantad** [comment]
The event looks like

[{'axes': {'x': 200, 'y': 0}, 'x': -1535.2, 'y': 0.0}, {'axes': {'x': 200, 'y': 1}, 'x': -1535.2, 'y': 0.0}, {'axes': {'x': 200, 'y': 2}, 'x': -1535.2, 'y': 0.0}, {'axes': {'x': 200, 'y': 3}, 'x': -1535.2, 'y': 0.0},

and it goes on until 'y' is 200 - This is bc I have 200 x positions. The x positions are constant bc that's how they had set it up for triggering in the example I mentioned.

Do I need to add num_time_points somewhere in my code? 
*I have not used the demo camera before, can you let me know where I can find that? Thank you.

**Comment by dpshepherd** [comment]
@kidantad - that code is from one of our scopes.

Before getting into the event structure, a few questions on the hardware.

Does the stage move when issue the SCAN command? 
Does the MS-2000 output the TTL signal when issued the SCAN command?
Is the camera actually in 'EXTERNAL' trigger mode? We've found we need to check the property to make sure it was actually set for the MM DCAM driver.

**Comment by kidantad** [comment]
Thank you for the updated code @dpshepherd This is my first time working with an external trigger, so I appreciate the help.

As far as hardware goes, the stage does move when I issue the SCAN command. I tried it first on the ASI console and then on pycromanager.

To see if the Camera is actually in "EXTERNAL' Trigger mode and if the MS2000 outputs a TTL signal, I checked by setting the camera settings to 'external', then I started the camera capture, which only showed a blank screen and then issued the SCAN command. When I do so, the camera takes an image once the SCAN command is issued. I did this on the ASI Console and the camera's software and not on pycromanager. Is there a better way to check if my controller is outputting a TTL signal?

**Comment by dpshepherd** [comment]
That all sounds good.

The next issue is that having the camera in EXTERNAL mode means it is expecting an external trigger for every frame.

Do you want the stage controller to stop and issue the camera a command at each position or perform continuous scan while the camera is free-running?

**Comment by kidantad** [comment]
Great. I want to perform a continuous scan while the camera is free running.

**Comment by dpshepherd** [comment]
Then you need to put the camera into START mode, so that it starts free-running upon an external trigger.

```
# set camera to START mode upon external input trigger
mmc.set_config('Camera-TriggerType','START')
mmc.wait_for_config('Camera-TriggerType','START')

# set camera to EXTERNAL trigger mode
mmc.set_config('Camera-TriggerSource','EXTERNAL')
mmc.wait_for_config('Camera-TriggerSource','EXTERNAL')
```

You should be able to set test this state using the camera software. If you don't have this setting in the DCAM software, that means your camera does not support this mode. I'm pretty sure all modern OrcaFlash cameras support the mode. I don't know what camera you are trying to use.

Once that works, then you can deal with the event structure.

**Comment by kidantad** [comment]
I am using OrcaFlash V4.0 and I do have the START setting on the DCAM software but it does not show up in micromanager. I was initially thinking of using the START mode as well, but I considered External edge because I want to eventually continuously scan in X, save the file, move up in Z and do it all over again. But I guess, that is also possible with the START mode if I give my camera the number of frames to take an image of while it's in the free-running mode correct?

**Comment by dpshepherd** [comment]
What version of the OrcaFlash v4.0 are you using?

**Comment by kidantad** [comment]
I am using Hamamatsu ORCA 4.0 V3.  I see, that makes sense.

**Comment by dpshepherd** [comment]
Page 38 of the [manual](https://www.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C13440-20CU_IM_En.pdf) explains External Edge mode

Page 44 of the [manual](https://www.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C13440-20CU_IM_En.pdf)explains Start trigger mode

This is why you only see one image. You'll need to get your hardware working correctly before dealing with events. If you are having trouble with the Micromanager DCAM driver, you'll need to contact Hamamatsu to troubleshoot. We have the `TriggerType` option in Micromanager present for both our OrcaFlash cameras (one 4.0v2 and one Fusion-BT).

**Comment by kidantad** [comment]
Okay, thank you very much. I will contact Hamamatsu about why the START trigger type is not showing up in micromanager for me. As far as events go, can I check back in with you in this thread once I have the hardware issue figured out? Assuming that I get START trigger type to work, I am guessing that I will need to include the number of timepoints in my event structure.

**Comment by dpshepherd** [comment]
Yes, not a problem. Good luck!

**Comment by kidantad** [comment]
Hello. I have solved the issue with the missing START trigger setting. For future reference, I had to change both the "Trigger" and "TRIGGER SOURCE" settings.

The stage is sending a signal and the camera receives and takes 200 images as I want it to. However, it is taking 200 images at one position even if my stage is moving the right amount. I'm guessing that is because I only have the start position in the event structure.  

This is a copy of the result from my event structure:
[{'axes': {'x': 200, 'y': 0}, 'x': 152.8, 'y': 0.0724}, {'axes': {'x': 200, 'y': 1}, 'x': 152.8, 'y': 0.0724}, {'axes': {'x': 200, 'y': 2}, 'x': 152.8, 'y': 0.0724} and so on for 200 positions.

Ideally my x positions will be my scan start to end

**Comment by henrypinkard** [comment]
It looks like you have the same xy positions in every event (i.e. `'x': 152.8, 'y': 0.0724`) Not sure why your stage would even move with these events. I would expect that you need to have this change from event to event

**Comment by kidantad** [comment]
I agree. The start and end positions are set with the SCANR and executed with the SCAN command. When I run the acquisition, I can see on my stage monitor that it is moving from the start to end position correctly . However, even after changing the event structure to a simple one like you suggested, the images still come out to be 200 images at one position. I am using beads as a sample to record this images and the images appear to be at one spot. 

The stage and camera are communicating well. If move 40umm instead of 20, the camera takes 400 images instead of 200 so I know that it is working good, but I don't understand why all 200 images appear to be in one position.

**Comment by kidantad** [comment]
I calculated the exposure time based on the Camera's setting. I used the Camera's software to check which exposure time at a reduced ROI will have the Camera run at 200fps. For my camera, it is 5ms for using half chip. What do you mean by calculating from the stage's setting?

**Comment by dpshepherd** [comment]
Because the stage sends a START signal when it passes the defined start point and that signal starts the camera in free running mode, synchronization is not guaranteed unless you match the exposure time of the camera to the stage motion. We do this calculation on [lines 146-155](https://github.com/QI2lab/OPM/blob/15101041435598f9892e7ed517cdbd8581726966/pycromanager-control/run_opm_stagescan.py#L146) for our specific hardware and goals.

You also have to calculate and test what range of stage speeds have no (or acceptable) motion blur for your specific hardware and goals. For example, we do not move the stage more than 400 nm between exposures for our application.

Running the microscope in this manner is somewhat involved and the details are highly dependent on the hardware and goals. Not sure I can give much more advice.

**Comment by kidantad** [comment]
I understand. Thank you very much. You have been very helpful and I really appreciate your help. I will look more into those calculations.
